File: sortList_java/sortList.java
public class sortList {

	public static ListNode sort(ListNode head) {
		if (head == null || head.next == null) {
			return head;
		}
		ListNode middle = head, tail = head, pre = head;
		while (tail != null && tail.next != null) {
			pre = middle;
			middle = middle.next;
			tail = tail.next.next;
		}
		pre.next = null;
		ListNode h1 = sort(head);
		ListNode h2 = sort(middle);
		return merge(h1, h2);
	}

	public static ListNode merge(ListNode h1, ListNode h2) {
		ListNode head, tmp;
		;
		if (h1 == null) {
			return h2;
		}
		if (h2 == null) {
			return h1;
		}
		if (h1.val > h2.val) {
			head = h2;
			tmp = h2;
			h2 = h2.next;
		} else {
			head = h1;
			tmp = h1;
			h1 = h1.next;
		}
		while (h1 != null && h2 != null) {
			if (h1.val > h2.val) {
				tmp.next = h2;
				tmp = tmp.next;
				h2 = h2.next;
			} else {
				tmp.next = h1;
				tmp = tmp.next;
				h1 = h1.next;
			}
		}
		if (h1 == null) {
			tmp.next = h2;
		} else {
			tmp.next = h1;
		}
		return head;
	}

	public static void main(String[] args) {

		ListNode a = new ListNode(8);
		a.next = new ListNode(4);
		 a.next.next =new ListNode(2);
		ListNode b = sort(a);

		while (b != null) {
			System.out.println(b.val);
			b = b.next;
		}

	}
}

File: sortlist.cpp
#include <iostream>
#include <assert.h>
using namespace std;
//Definition for singly-linked list.
 struct ListNode {
     int val;
     ListNode *next;
     ListNode(int x) : val(x), next(NULL) {}
 };


 class Solution{
 public:
	 ListNode* sortList(ListNode* head){
	 return divide(head);
	 }


	 ListNode* divide(ListNode* head){
		 if(head == NULL || head->next == NULL){
			return head;
		 }
		 ListNode *middle=head,*tail=head,*pre;
		 while(tail !=NULL && tail->next != NULL){
			 pre=middle;
			 middle=middle->next ;
			 tail=tail->next->next; 
		 }
		 pre->next = NULL;
		 ListNode *first=divide(head);
		 ListNode *second=divide(middle);
		 return merge(first,second);
	 }

	 ListNode *merge(ListNode *first,ListNode *second){
		ListNode *head,*list;
		if(first != NULL && second != NULL){
			if(first->val < second->val ){
			 head = first;
			 first = first->next ;
			}else{
			 head = second;
			 second = second->next ;
			}
			list=head;
		 }

		while(first != NULL && second != NULL){
			if(first->val < second->val ){
			 list->next = first;
			 first = first->next ;
			}else{
			 list->next = second ;
			 second = second->next ;
			}
			list=list->next ;
		}

		if(first == NULL){
		    list->next = second;
		}else{
			list->next =first;
		}
		return head;
	 }
 };


 int main(){
 ListNode *head=new ListNode(1),*tmp=head;
 head->next=new ListNode(8);
 head->next->next=new ListNode(5);
 Solution s;
 s.sortList(head);
 while(head != NULL) {
	cout << head->val << "  " << endl;
	head=head->next ;
 }

 return 0;
} 
